libsolidity/experimental/ast/TypeSystem.cpp

```diff
@@ -302,7 +302,7 @@ experimental::Type TypeSystem::type(TypeConstructor _constructor, std::vector<Ty
 	return TypeConstant{_constructor, _arguments};
 }
 
-experimental::Type TypeEnvironment::fresh(Type _type)
+experimental::Type TypeEnvironment::resolveWithFreshTypeVariables(Type _type)
```

Comment:
This is a *really* bad name change IMHO :-).

Reply:
In other words: please change it back ;-).

Reply:
Can we at least have something more self-explanatory than just `fresh()`. It's way too vague. I was already bitten by was that it resolves the type which I completely did not expect from the current name.

Reply:
I removed the fixup so it's back to what it was originally, but I still think it needs a better name for this. It `fresh()` that seems bad to me.

Reply:
You can make it ``freshen``, if you like, but that's the only thing that happens there really...
Mentioning the fact that this involves "resolving" is like for an assignment ``a = b`` saying "use the value of ``b``, but really the last assigned value of ``b``, and also if ``b`` was assigned to ``c`` before, make sure to have used the latest value of ``c`` to earlier" - that's just noise :-).

Reply:
So as a general rule: it's pretty much never valid to use a type without resolving it - it's just some cases that you know that they're already resolved or in which you know that you only need the outermost layer, so you don't need to know recursively. But it's an implementation detail, really, it's just how you fetch the actual values of type expressions.

Reply:
> So as a general rule: it's pretty much never valid to use a type without resolving it

Well, I'm slowly realizing that I'm supposed to resolve all the time. It seemed like a relatively heavy operation and something you should be aware of when it happens. Some of the bugs I've been fighting with recently came down to the type not being resolved, so I guess the way you see it makes sense.

I still think that `fresh()` or `freshen()` is too vague though. I'd make it `refreshGenericTypeVariables()` at least. After #14606 we have the other type of variables that does not get refreshed.

> On the other hand, this should become less pronounced anyways after https://github.com/ethereum/solidity/pull/14606 I would assume

Actually, yes. The only use of it I'm left with is inside `polymorphicInstance()`, and that will probably go away eventually too. The rest I have replaced with my function for generalizing the variables. And anyway, the only difference between `fresh()` and and my generalization is which variables are replaced (generic of fixed ones) so I guess I'll end up replacing them both with a more generic function doing substitution.

Reply:
Yeah, the very idea of https://github.com/ethereum/solidity/pull/14606 is pretty much to replace most (if not all) instances of this kind of freshening by generalization of types with fixed type variables to exactly avoid forgetting to freshen - that includes, in particular ``polymorphicInstance``, for which the name actually becomes more fitting, since it will actually give you a polymorphic instance of a fixed type by generalization.